StatsD connections should control the endpoint

trevnorris opened the pull request:
The main change here is for each connection to control its own reconnection logic. That'll make it easier to abstract the implementation out for a virtual class.

trevnorris committed:
agents: remove need for loop on create()

Not going to worry about the overhead of needing to call uv_loop_init()
on every call to create() since it's almost never called. This way we
don't need to pass in a uv_loop_t on the call.

If the overhead is a problem in the future then we'll change it so
there's a StatsDEndpoint loop just for that call and add a lock for
access so it's thread-safe.

PR-URL: https://github.com/nodesource/nsolid/pull/121
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

trevnorris committed:
agents: only create on correct protocol

Only tcp and udp are supported, so return early if the protocol doesn't
match.

PR-URL: https://github.com/nodesource/nsolid/pull/121
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

trevnorris committed:
agents: have tcp/udp retry in statsd

Instead of handling the retry logic in the StatsDAgent, have each
connection handle its own endpoint. This way we can abstract the
connection class with a future virtual class and allow multiple
instances of StatsDAgent to be instantiated.

This change also involves other cleanup, and changing the way the timer
works so it retries to connect to the StatsD endpoint, after all IP
addresses have been tried, every 3 seconds.

PR-URL: https://github.com/nodesource/nsolid/pull/121
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -697,9 +697,10 @@ int StatsDAgent::config_handles() {
+    } else if (endpoint_->protocol() == "udp") {
       setup_udp();
+    } else {
+      abort();  // Unreachable
```
Why this change? IIUC the current code behaves the same but the assertion message will give print the wrong protocol

trevnorris replied:
The next PR will abstract away `tcp_` and `udp_` to a single virtual class stored by the `StatsDAgent`. Do we want to put a `Debug()` message before this so it prints the endpoint information?

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -997,9 +998,10 @@ void StatsDAgent::retry_timer_cb_(nsuv::ns_timer*, WeakStatsDAgent agent_wp) {
+  } else if (endpoint_->protocol() == "udp") {
     agent->setup_udp();
+  } else {
+    abort();  // Unreachable
```
Ditto

trevnorris replied:
Same here about adding the `Debug()` message.

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -39,21 +39,28 @@ inline void DebugJSON(const char* str, const json& msg) {
   addr_str_lock_.destroy();
+  // Make sure close_and_delete() is always called before the destructor is
+  // called.
+  ASSERT_NULL(retry_timer_);
```
Should we also check for `tcp_` to be NULL?

trevnorris replied:
Sure.

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -118,6 +106,26 @@ int StatsDTcp::write(const string_vector& messages) {
+void StatsDTcp::setup(StatsDEndpoint* endpoint) {
+  ASSERT_NOT_NULL(endpoint);
+
+  int r = retry_timer_->stop();
+  if (r) {
+    // Nothing to do.
+  }
```
We should add a `USE()` macro, or maybe just use the one from node (though it's in internals)

trevnorris replied:
I forgot that was a thing. I'll use that.

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -118,6 +106,26 @@ int StatsDTcp::write(const string_vector& messages) {
+  }
+  if (tcp_ != nullptr) {
+    tcp_->close_and_delete();
+    tcp_ = nullptr;
```
no need to set tcp_ to null here

trevnorris replied:
will do

santigimeno commented on agents/statsd/src/statsd_agent.h:
```diff
@@ -34,8 +36,13 @@ static const nlohmann::json default_agent_config = {
-class StatsDTcp {
+using WeakStatsDTcp = std::weak_ptr<StatsDTcp>;
+
+class StatsDTcp : public std::enable_shared_from_this<StatsDTcp> {
```
IIUC the [documentation](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this), If you're inheriting from `std::enable_shared_from_this` the `StatsDTcp` instance should be stored in a `shared_ptr`.

trevnorris replied:
Looking over it more carefully, I realize now that it's not necessary to inherit from `std::enable_shared_from_this` if the life of the timer is controlled more carefully. Going to do the rewrite.

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -280,6 +325,25 @@ int StatsDUdp::write(const string_vector& messages) {
+  if (r) {
+    // Nothing to do.
+  }
+  if (udp_ != nullptr) {
+    udp_->close_and_delete();
+    udp_ = nullptr;
+  }
+  udp_ = new nsuv::ns_udp();
```
```suggestion
  if (udp_ != nullptr) {
    udp_->close_and_delete();
  }
  udp_ = new nsuv::ns_udp();
```

trevnorris replied:
will do

santigimeno reviewed (changes requested):
Mostly LGTM. Left some comments

trevnorris committed:
agents: remove Disconnected

Now that connect_() is immediately called after a disconnection, that
status is never actually used anymore.

PR-URL: https://github.com/nodesource/nsolid/pull/121
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

santigimeno commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -161,18 +164,18 @@ void StatsDTcp::connect_cb_(nsuv::ns_connect<nsuv::ns_tcp>* req,
     return;
   }
-  tcp->connect_();
+
+  Debug("Error connecting to all addresses. Retry in 3 seconds...\n");
+  tcp->status(ConnectionError);
+  // Reattempt all IP addresses after 3 seconds.
+  tcp->addr_index_ = 0;
+  tcp->prep_retry_timer_();
+  ASSERT_EQ(0, tcp->retry_timer_->start(
+        StatsDTcp::retry_timer_cb_, 3000, 0, tcp));
```
Nit: can this logic refactor into a helper function ?

trevnorris replied:
will do
